tachyon/crypto/commitments/fri/two_adic_fri_pcs.h

```diff
@@ -0,0 +1,403 @@
+          math::RowMajorMatrix<F> block =
+              mat.topRows(num_rows >> fri_.log_blowup);
+          ReverseMatrixIndexBits(block);
+          std::vector<ExtF> ys = InterpolateCoset(
+              std::move(block),
+              F::FromMontgomery(F::Config::kSubgroupGenerator), point);
+          const ExtF alpha_pow_offset = alpha.Pow(num_reduced[log_num_rows]);
+          const std::pmr::vector<ExtF> alpha_powers =
+              ExtF::GetSuccessivePowers(num_cols, alpha);
+          ExtF reduced_ys = ExtF::Zero();
+          for (size_t c = 0; c < num_cols; ++c) {
+            reduced_ys += alpha_powers[c] * ys[c];
+          }
+          std::vector<ExtF> reduced_rows = DotExtPowers(mat, alpha);
+          std::vector<ExtF>& inv_denom = inv_denoms[point];
+          for (size_t i = 0; i < reduced_opening_for_log_num_rows.size(); ++i) {
+            reduced_opening_for_log_num_rows[i] +=
+                alpha_pow_offset * (reduced_rows[i] - reduced_ys) *
+                inv_denom[i];
+          }
+          num_reduced[log_num_rows] += num_cols;
+          opened_values_for_mat[point_idx] = ys;
+        }
+        opened_values_for_round[matrix_idx] = opened_values_for_mat;
+      }
+      opened_values[round] = opened_values_for_round;
+    }
+    std::vector<std::vector<ExtF>> fri_input;
+    fri_input.reserve(reduced_openings.size() - 1);
+    for (size_t i = reduced_openings.size() - 1; i != SIZE_MAX; --i) {
+      if (!reduced_openings[i].empty()) {
+        fri_input.push_back(std::move(reduced_openings[i]));
+      }
+    }
+
+    TwoAdicFriProof fri_proof = TwoAdicFriPcsProve<InputMMCS>(
+        fri_, std::move(fri_input), challenger,
+        [this, log_global_max_num_rows, &prover_data_by_round](size_t index) {
+          size_t num_rounds = prover_data_by_round.size();
+          std::vector<BatchOpening<InputMMCS>> ret(num_rounds);
+          OPENMP_PARALLEL_FOR(size_t round = 0; round < num_rounds; ++round) {
+            Proof proof;
+            std::vector<std::vector<F>> openings;
+            const ProverData& prover_data = prover_data_by_round[round];
+            size_t log_max_num_rows =
+                base::bits::CheckedLog2(mmcs_.GetMaxRowSize(prover_data));
+            size_t bits_reduced = log_global_max_num_rows - log_max_num_rows;
+            size_t reduced_index = index >> bits_reduced;
+            CHECK(mmcs_.CreateOpeningProof(reduced_index, prover_data,
+                                           &openings, &proof));
+            ret[round] = BatchOpening<InputMMCS>{openings, proof};
+          }
+          return ret;
+        });
+    *openings = opened_values;
+    *proof = fri_proof;
+    return true;
+  }
+
+  [[nodiscard]] bool VerifyOpeningProof(
+      const std::vector<Commitment>& commits_by_round,
+      const std::vector<std::vector<Coset>>& domains_by_round,
+      const std::vector<
+          std::vector<std::vector<std::tuple<ExtF, std::vector<ExtF>>>>>&
```

Comment:
While the tuple can be separated into two different containers, it may be worth learning more about the whole process and determining how to deal with this later. This would also be better since the first `ExtF` of the container in the unittest is a constant, so the table could be separated into a constant and a container, but in reality it may be varying values. I'll just put a comment here to remove the `std::tuple` in the future.